xtask/src/ci/spec_links.rs

```diff
@@ -0,0 +1,103 @@
+        // URL is `../rx.x.x.html`
+        s.starts_with("r0.6.1") || s.starts_with("unstable#")
+    }),
+    ("https://matrix.org/docs/spec/server_server/", |s| {
+        s.starts_with("r0.1.4") || s.starts_with("unstable#")
+    }),
+    ("https://matrix.org/docs/spec/application_service/", |s| {
+        s.starts_with("r0.1.2") || s.starts_with("unstable#")
+    }),
+    ("https://matrix.org/docs/spec/identity_service/", |s| {
+        s.starts_with("r0.3.0") || s.starts_with("unstable#")
+    }),
+    ("https://matrix.org/docs/spec/push_gateway/", |s| {
+        s.starts_with("r0.1.1") || s.starts_with("unstable#")
+    }),
+];
+
+pub(crate) fn check_spec_links(path: &Path) -> Result<()> {
+    // This is WAY overkill but since there are a few mixed in ruma-common
+    // and this would catch any wrong version anywhere it's probably ok
+    for (split, version_fn) in SPLITS {
+        walk_dirs(path, split, *version_fn)?;
+    }
+    Ok(())
+}
+
+fn walk_dirs(path: &Path, split: &str, version_match: fn(&str) -> bool) -> Result<()> {
+    if path.is_dir() {
+        for entry in fs::read_dir(path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                walk_dirs(&path, split, version_match)?;
+            } else {
+                let content = fs::read_to_string(&path)?;
+                let split = content.split(split).collect::<Vec<_>>();
+
+                match split.as_slice() {
+                    // No spec link was found
+                    // `[]` is only to satisfy exhaustiveness checking
+                    [] | [_] => {}
+                    [pre, post, rest @ ..] => {
+                        if pre.lines().rev().next().map_or(false, |l| l.starts_with("//!"))
+                            && !version_match(post)
+                        {
+                            return Err(format!(
+                                "error: spec URL with no version number\nfrom: {}\n{}",
+                                path.display(),
+                                &post[0..25]
+                            )
+                            .into());
+                        }
+
+                        check_rest(rest, &path, version_match)?;
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn check_rest(rest: &[&str], path: &Path, version_match: fn(&str) -> bool) -> Result<()> {
+    match rest {
+        [] => Ok(()),
+        [rest] => {
+            if !version_match(rest) {
+                Err(format!(
+                    "error: spec URL with no version number\nfrom: {}\n{}",
+                    path.display(),
+                    &rest[0..25]
+                )
+                .into())
+            } else {
+                Ok(())
+            }
+        }
+        [pre, post, rest @ ..] => {
+            if pre.lines().rev().next().map_or(false, |l| l.starts_with("//!"))
```

Comment:
Same as above.

Reply:
Do you think the check makes sense even? Should we be checking any link in any file even if it's in regular comments?

Reply:
Oh. Yeah actually we should probably be checking everything. At the very least `///` should also be included, but `//` comments: Why not 🤷🏼